data-prepper-plugins/newline-codecs/src/main/java/org/opensearch/dataprepper/plugins/codec/newline/NewlineDelimitedOutputCodec.java

```diff
@@ -4,42 +4,89 @@
 import org.opensearch.dataprepper.model.codec.OutputCodec;
 import org.opensearch.dataprepper.model.event.Event;
 
 import java.io.IOException;
 import java.io.OutputStream;
+import java.util.Map;
+import java.util.Objects;
 
 /**
  * An implementation of {@link OutputCodec} which deserializes Data-Prepper events
  * and writes them to Output Stream as ND-JSON data
  */
-@DataPrepperPlugin(name = "newline", pluginType = OutputCodec.class)
+@DataPrepperPlugin(name = "newline", pluginType = OutputCodec.class, pluginConfigurationType = NewlineDelimitedOutputConfig.class)
 public class NewlineDelimitedOutputCodec implements OutputCodec {
+    private static final String NDJSON = "ndjson";
+    private static final String MESSAGE_FIELD_NAME = "message";
+    private final String headerDestination;
+    private static final ObjectMapper objectMapper = new ObjectMapper();
+    private final NewlineDelimitedOutputConfig config;
+
+    @DataPrepperPluginConstructor
+    public NewlineDelimitedOutputCodec(final NewlineDelimitedOutputConfig config) {
+        Objects.requireNonNull(config);
+        this.config = config;
+        headerDestination = config.getHeaderDestination();
+    }
 
     @Override
     public void start(final OutputStream outputStream) throws IOException {
-        // TODO: implement
+        Objects.requireNonNull(outputStream);
     }
 
     @Override
     public void writeEvent(final Event event, final OutputStream outputStream) throws IOException {
-        // TODO: get the event data and
-        //  get the header record and message record and write event data to the outputstream
+        Objects.requireNonNull(event);
+        final Map<String, Object> eventDataMap = event.toMap();
+        if (eventDataMap.containsKey(headerDestination) || eventDataMap.containsKey(MESSAGE_FIELD_NAME)) {
+            final Object headerValue = eventDataMap.get(headerDestination);
+            if (headerValue != null) {
+                writeToOutputStream(outputStream, headerValue);
+                eventDataMap.remove(headerDestination);
+            }
+
+            final Object value = eventDataMap.get(MESSAGE_FIELD_NAME);
+            if (value != null) {
+                writeToOutputStream(outputStream, value);
+            }
+        } else {
+            writeToOutputStream(outputStream, eventDataMap);
+        }
     }
 
     @Override
     public void complete(final OutputStream outputStream) throws IOException {
-        // TODO: Close the output stream
+        outputStream.close();
     }
 
-    private void writeByteArrayToOutputStream(final OutputStream outputStream, final Object object) throws IOException {
-        // TODO: common method to write byte array data to OutputStream
+    private void writeToOutputStream(final OutputStream outputStream, final Object object) throws IOException {
+        boolean isExcludeKeyAvailable = !Objects.isNull(config.getExcludeKeys());
```

Comment:
Nit: Objects.nonNull(config.getExcludeKeys);

Reply:
I would recommend defaulting `excludeKeys` to be an empty list. That way you don't need this check, and the code below can just be

```
                for (String key : config.getExcludeKeys()) {
                    if (map.containsKey(key)) {
                        map.remove(key);
                    }
                }
```

without the check for if isExcludeKeyAvailable

Reply:
Done.!